File: go/155MinStack.go
package main

import (
	"container/list"
	"fmt"
)

type MinStack struct {
	min  int
	data *list.List
}

/** initialize your data structure here. */
func Constructor() MinStack {
	return MinStack{min: 0, data: list.New()}
}

func (this *MinStack) Push(x int) {
	if this.data.Len() == 0 {
		this.data.PushFront(x)
		this.min = x
	} else {
		if this.min >= x {
			this.data.PushFront(this.min) // cache the second minimum number, this is the key to retrieve min in constant time
			this.min = x
		}
		this.data.PushFront(x)
	}
}

func (this *MinStack) Pop() {
	if this.data.Front().Value.(int) == this.min {
		this.data.Remove(this.data.Front())
		if this.data.Front() != nil {
			this.min = this.data.Front().Value.(int)
			this.data.Remove(this.data.Front())
		}
	} else {
		this.data.Remove(this.data.Front())
	}
}

func (this *MinStack) Top() int {
	return this.data.Front().Value.(int)
}

func (this *MinStack) GetMin() int {
	return this.min
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 ["MinStack","push","push","push","top","pop","getMin","pop","getMin","pop","push","top","getMin","push","top","getMin","pop","getMin"]
[[],[2147483646],[2147483646],[2147483647],[],[],[],[],[],[],[2147483647],[],[],[-2147483648],[],[],[],[]]
*/

func main() {
	obj := Constructor()
	obj.Push(2147483646)
	obj.Push(2147483646)
	obj.Push(2147483647)
	fmt.Println("top", obj.Top())
	obj.Pop()
	fmt.Println("min", obj.GetMin())
	obj.Pop()
	fmt.Println("min", obj.GetMin())
	obj.Pop()
	obj.Push(2147483647)
	fmt.Println("top", obj.Top())
	fmt.Println("min", obj.GetMin())
	obj.Push(-2147483648)
	fmt.Println("top", obj.Top())
	fmt.Println("min", obj.GetMin())
	obj.Pop()
	fmt.Println("min", obj.GetMin())

}

File: cpp/155MinStack.cpp
#include <iostream>
#include <set>
#include <vector>
#include <iterator>
#include <utility>
#include <cstdlib>
#include <limits>
#include <functional>
#include <stack>
#include <unordered_set>
#include <unordered_map>
#include <queue>
#include <algorithm>
#include <cstdint>

using namespace std;

template<class Container>
void printContainer(const Container& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }

    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[";
        }
        cout << *it;
        if (it == prev(container.end())) {
            cout << "]" << endl;
        } else {
            cout << ", ";
        }
    }
}

template<class T>
void print2DimVec(const vector<vector<T>>& container) {
//    return;

    if (container.size() == 0) {
        cout << "[]" << endl;
        return;
    }
    for (auto it = container.begin(); it != container.end(); it++) {
        if (it == container.begin()) {
            cout << "[" << endl;
        }
        printContainer(*it);
        if (it == prev(container.end())) {
            cout << "]" << endl;
        }
    }
}

class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() :
        m_min(0) {

    }

    void push(int x) {
        if (m_stack.size() == 0) {
            m_stack.push(x);
            m_min = x;
        } else {
            if (m_min >= x) {
                m_stack.push(m_min);
                m_min = x;
            }
            m_stack.push(x);
        }
    }

    void pop() {
        if (top() == m_min) {
            m_stack.pop();
            if (!m_stack.empty()) {
                m_min = m_stack.top();
                m_stack.pop();
            }
        } else {
            m_stack.pop();
        }
    }

    int top() {
        return m_stack.top();
    }

    int getMin() {
        return m_min;
    }
private:
    int m_min;
    stack<int> m_stack;
};

int main(){
    MinStack* obj = new MinStack();
    obj->push(3);
    cout << "top" << obj->top() << endl;
    obj->push(2);
    cout << "min" << obj->getMin() << endl;
    obj->push(4);
    cout << "min" << obj->getMin() << endl;
    obj->pop();
    cout << "min" << obj->getMin() << endl;
    obj->pop();
    cout << "min" << obj->getMin() << endl;

    cout << "*************output*************" << endl;

    return 0;
}

